### PR TITLE
API: check locator and formatter args when passed

### LINKED ISSUE #10769: DOC: set_major_locator could check that its getting a Locator (was EngFormatter broken?)
### Bug report

`matplotlib.ticker.EngFormatter` seems to be broken:

```
f, ax = plt.subplots(1, 1, figsize=(15, 10))
plt.plot([10**4, 11*10**4], [100, 150])
ax.xaxis.set_major_locator(matplotlib.ticker.EngFormatter())
```
```
Error in callback <function install_repl_displayhook.<locals>.post_execute at 0x7f60a9b120d0> (for post_execute):

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.5/dist-packages/matplotlib/pyplot.py in post_execute()
    147             def post_execute():
    148                 if matplotlib.is_interactive():
--> 149                     draw_all()
    150 
    151             # IPython >= 2

/usr/local/lib/python3.5/dist-packages/matplotlib/_pylab_helpers.py in draw_all(cls, force)
    134         for f_mgr in cls.get_all_fig_managers():
    135             if force or f_mgr.canvas.figure.stale:
--> 136                 f_mgr.canvas.draw_idle()
    137 
    138 atexit.register(Gcf.destroy_all)

/usr/local/lib/python3.5/dist-packages/matplotlib/backend_bases.py in draw_idle(self, *args, **kwargs)
   2053         if not self._is_idle_drawing:
   2054             with self._idle_draw_cntx():
-> 2055                 self.draw(*args, **kwargs)
   2056 
   2057     def draw_cursor(self, event):

/usr/local/lib/python3.5/dist-packages/matplotlib/backends/backend_agg.py in draw(self)
    429             # if toolbar:
    430             #     toolbar.set_cursor(cursors.WAIT)
--> 431             self.figure.draw(self.renderer)
    432             # A GUI class may be need to update a window using this draw, so
    433             # don't forget to call the superclass.

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/figure.py in draw(self, renderer)
   1473 
   1474             mimage._draw_list_compositing_images(
-> 1475                 renderer, self, artists, self.suppressComposite)
   1476 
   1477             renderer.close_group('figure')

/usr/local/lib/python3.5/dist-packages/matplotlib/image.py in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    139     if not_composite or not has_images:
    140         for a in artists:
--> 141             a.draw(renderer)
    142     else:
    143         # Composite any adjacent images together

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/axes/_base.py in draw(self, renderer, inframe)
   2605             renderer.stop_rasterizing()
   2606 
-> 2607         mimage._draw_list_compositing_images(renderer, self, artists)
   2608 
   2609         renderer.close_group('axes')

/usr/local/lib/python3.5/dist-packages/matplotlib/image.py in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    139     if not_composite or not has_images:
    140         for a in artists:
--> 141             a.draw(renderer)
    142     else:
    143         # Composite any adjacent images together

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in draw(self, renderer, *args, **kwargs)
   1185         renderer.open_group(__name__)
   1186 
-> 1187         ticks_to_draw = self._update_ticks(renderer)
   1188         ticklabelBoxes, ticklabelBoxes2 = self._get_tick_bboxes(ticks_to_draw,
   1189                                                                 renderer)

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in _update_ticks(self, renderer)
   1023 
   1024         interval = self.get_view_interval()
-> 1025         tick_tups = list(self.iter_ticks())  # iter_ticks calls the locator
   1026         if self._smart_bounds and tick_tups:
   1027             # handle inverted limits

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in iter_ticks(self)
    966         Iterate through all of the major and minor ticks.
    967         """
--> 968         majorLocs = self.major.locator()
    969         majorTicks = self.get_major_ticks(len(majorLocs))
    970         self.major.formatter.set_locs(majorLocs)

TypeError: __call__() missing 1 required positional argument: 'x'

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/usr/local/lib/python3.5/dist-packages/IPython/core/formatters.py in __call__(self, obj)
    339                 pass
    340             else:
--> 341                 return printer(obj)
    342             # Finally look for special method names
    343             method = get_real_method(obj, self.print_method)

/usr/local/lib/python3.5/dist-packages/IPython/core/pylabtools.py in <lambda>(fig)
    236 
    237     if 'png' in formats:
--> 238         png_formatter.for_type(Figure, lambda fig: print_figure(fig, 'png', **kwargs))
    239     if 'retina' in formats or 'png2x' in formats:
    240         png_formatter.for_type(Figure, lambda fig: retina_figure(fig, **kwargs))

/usr/local/lib/python3.5/dist-packages/IPython/core/pylabtools.py in print_figure(fig, fmt, bbox_inches, **kwargs)
    120 
    121     bytes_io = BytesIO()
--> 122     fig.canvas.print_figure(bytes_io, **kw)
    123     data = bytes_io.getvalue()
    124     if fmt == 'svg':

/usr/local/lib/python3.5/dist-packages/matplotlib/backend_bases.py in print_figure(self, filename, dpi, facecolor, edgecolor, orientation, format, **kwargs)
   2208                     orientation=orientation,
   2209                     dryrun=True,
-> 2210                     **kwargs)
   2211                 renderer = self.figure._cachedRenderer
   2212                 bbox_inches = self.figure.get_tightbbox(renderer)

/usr/local/lib/python3.5/dist-packages/matplotlib/backends/backend_agg.py in print_png(self, filename_or_obj, *args, **kwargs)
    509 
    510     def print_png(self, filename_or_obj, *args, **kwargs):
--> 511         FigureCanvasAgg.draw(self)
    512         renderer = self.get_renderer()
    513         original_dpi = renderer.dpi

/usr/local/lib/python3.5/dist-packages/matplotlib/backends/backend_agg.py in draw(self)
    429             # if toolbar:
    430             #     toolbar.set_cursor(cursors.WAIT)
--> 431             self.figure.draw(self.renderer)
    432             # A GUI class may be need to update a window using this draw, so
    433             # don't forget to call the superclass.

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/figure.py in draw(self, renderer)
   1473 
   1474             mimage._draw_list_compositing_images(
-> 1475                 renderer, self, artists, self.suppressComposite)
   1476 
   1477             renderer.close_group('figure')

/usr/local/lib/python3.5/dist-packages/matplotlib/image.py in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    139     if not_composite or not has_images:
    140         for a in artists:
--> 141             a.draw(renderer)
    142     else:
    143         # Composite any adjacent images together

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/axes/_base.py in draw(self, renderer, inframe)
   2605             renderer.stop_rasterizing()
   2606 
-> 2607         mimage._draw_list_compositing_images(renderer, self, artists)
   2608 
   2609         renderer.close_group('axes')

/usr/local/lib/python3.5/dist-packages/matplotlib/image.py in _draw_list_compositing_images(renderer, parent, artists, suppress_composite)
    139     if not_composite or not has_images:
    140         for a in artists:
--> 141             a.draw(renderer)
    142     else:
    143         # Composite any adjacent images together

/usr/local/lib/python3.5/dist-packages/matplotlib/artist.py in draw_wrapper(artist, renderer, *args, **kwargs)
     53                 renderer.start_filter()
     54 
---> 55             return draw(artist, renderer, *args, **kwargs)
     56         finally:
     57             if artist.get_agg_filter() is not None:

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in draw(self, renderer, *args, **kwargs)
   1185         renderer.open_group(__name__)
   1186 
-> 1187         ticks_to_draw = self._update_ticks(renderer)
   1188         ticklabelBoxes, ticklabelBoxes2 = self._get_tick_bboxes(ticks_to_draw,
   1189                                                                 renderer)

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in _update_ticks(self, renderer)
   1023 
   1024         interval = self.get_view_interval()
-> 1025         tick_tups = list(self.iter_ticks())  # iter_ticks calls the locator
   1026         if self._smart_bounds and tick_tups:
   1027             # handle inverted limits

/usr/local/lib/python3.5/dist-packages/matplotlib/axis.py in iter_ticks(self)
    966         Iterate through all of the major and minor ticks.
    967         """
--> 968         majorLocs = self.major.locator()
    969         majorTicks = self.get_major_ticks(len(majorLocs))
    970         self.major.formatter.set_locs(majorLocs)

TypeError: __call__() missing 1 required positional argument: 'x'

<Figure size 1080x720 with 1 Axes>
```

**Matplotlib version**

  * Operating system: Ubuntu 16.04
  * Matplotlib version: 2.2.0
  * Matplotlib backend (`print(matplotlib.get_backend())`): module://ipykernel.pylab.backend_inline
  * Python version: 3.5.2 (IPython 6.2.1)
  * Jupyter version (if applicable): Jupyter Lab beta 0.31.8

### PR DESCRIPTION
**Pull request by jklymak**
## PR Summary

Right now, `ax.set_major_formatter` and `ax.set_major_locator` throw pretty deep error messages if the wrong type of data is passed to them.  This catches at entry. 

Closes #10769

```python
import matplotlib
import matplotlib.pyplot as plt

f, ax = plt.subplots(1, 1, figsize=(15, 10))
plt.plot([10**4, 11*10**4], [100, 150])
ax.xaxis.set_major_formatter(matplotlib.ticker.LogLocator())
plt.show()
```
Now returns:
```
Traceback (most recent call last):
  File "testEngTicker.py", line 6, in <module>
    ax.xaxis.set_major_formatter(matplotlib.ticker.LogLocator())
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 1573, in set_major_formatter
    "formatter argument must be matplotlib.ticker.Formatter")
ValueError: formatter argument must be matplotlib.ticker.Formatter
```
instead of something down in the bowels of drawing the ticker...

## PR Checklist

- [x] Code is PEP 8 compliant

### CODE REVIEW

**Comment by afvincent** [comment]
@jklymak For what reason did you labelled it with the “API consistency” tag?

**Comment by jklymak** [comment]
Umm, because there is no argument-checking flag.

**Comment by jklymak** [comment]
There!

**Comment by afvincent** [comment]
Well it was more by curiosity that I was asking ^^. The docstrings seemed to be pretty clear about the kind of expected inputs and your changes is simply enforcing that information, is it not ;)?

**Comment by jklymak** [comment]
Well, the documentation is specifying the API 😉

**Comment by anntzer** [comment]
I don't really like the change, in general we are (and Python is) pretty lax in accepting anything that duck types correctly, and just let things blow up if the argument has the wrong type.

Regarding the error message (if this goes in): an instance of a subclass of Ticker is also an instance of Ticker (literally, as returned by `isinstance`...) so I wouldn't try to make it longer.

**Comment by jklymak** [comment]
Well, I'm certainly willing to go by convention, but the un-patched error is in `draw()`, so its a bit hopeless for the uninitiated to know that the error was actually in `set_major_formatter`:

```
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/lib/matplotlib/backends/backend_qt5.py", line 520, in _draw_idle
    self.draw()
  File "/Users/jklymak/matplotlib/lib/matplotlib/backends/backend_agg.py", line 426, in draw
    self.figure.draw(self.renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/figure.py", line 1504, in draw
    renderer, self, artists, self.suppressComposite)
  File "/Users/jklymak/matplotlib/lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 2588, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "/Users/jklymak/matplotlib/lib/matplotlib/image.py", line 139, in _draw_list_compositing_images
    a.draw(renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/artist.py", line 55, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 1189, in draw
    ticks_to_draw = self._update_ticks(renderer)
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 1027, in _update_ticks
    tick_tups = list(self.iter_ticks())  # iter_ticks calls the locator
  File "/Users/jklymak/matplotlib/lib/matplotlib/axis.py", line 973, in iter_ticks
    self.major.formatter.set_locs(majorLocs)
AttributeError: 'LogLocator' object has no attribute 'set_locs'
Traceback (most recent call last):
  File "/Users/jklymak/matplotlib/lib/matplotlib/axes/_base.py", line 3627, in format_xdata
    return self.fmt_xdata(x)
TypeError: 'NoneType' object is not callable
```

Worse, if you happen to drag your mouse across the figure window, the same error is emitted repeatedly in each draw attempt.

**Comment by jklymak** [comment]
If it looks like this will get approved, I'll make some tests so codecov is happy.  But I'm not advocating strongly that it should be merged if its really atypical to have this type of argument checking...

**Comment by anntzer** [comment]
(I'm not trying to block the change, but I think it's worth getting some more opinions...)

**Comment by story645** [comment]
@jklymak I'm +1 on this if it has tests and want this stuff to be more common in the codebase.

**Comment by jklymak** [comment]
Added tests....  happy to be tutored on making these parameterized...

**Comment by efiring** [comment]
The cost of putting in these checks is low; they are simple and fast.  I'm not in favor of checking all inputs at every stage, but this looks like a case where the cost/benefit ratio is favorable, so I lean towards accepting this PR.

File: lib/matplotlib/axis.py [hunk 1]
```diff
@@ -1568,6 +1568,9 @@ def set_major_formatter(self, formatter):
 
         ACCEPTS: A :class:`~matplotlib.ticker.Formatter` instance
         """
+        if not isinstance(formatter, mticker.Formatter):
```
Comment (timhoffm) [inline lib/matplotlib/axis.py]: To support duck typing, Python code tries to avoid `isinstance(obj, cls)` in favor of `hasattr(obj, name)`. Not sure about the canonical attributes of methods of locators and formatters, however, I would suggest checks like `if not hasattr(formatter, 'format_data')` and `if not hasattr(locator, 'tick_values')`

Comment (jklymak) [reply lib/matplotlib/axis.py]: Does that become an AttributeError in that case?

Comment (jklymak) [reply lib/matplotlib/axis.py]: eh, and probably should have been TypeError if not and AttributeError.  Certainly ValueError is incorrect?

**Comment by timhoffm** [comment]
Parametrizing would be something like:

~~~
@pytest.mark.parametrize('setter', [
    ax.xaxis.set_minor_formatter,
    ax.xaxis.set_major_formatter,
])
def test_set_formatter_type(setter):
    fig, ax = plt.subplots()
    with pytest.raises(ValueError):
        setter(matplotlib.ticker.LogFormatter())
~~~

or 

~~~
@pytest.mark.parametrize('setter,cls', [
    (ax.xaxis.set_minor_formatter, matplotlib.ticker.LogFormatter),
    (ax.xaxis.set_major_formatter, matplotlib.ticker.LogFormatter),
    (ax.xaxis.set_minor_locator, matplotlib.ticker.LogLocator),
    (ax.xaxis.set_major_locator, matplotlib.ticker.LogLocator),
])
def test_set_formatter_or_locator_type(setter, cls):
    fig, ax = plt.subplots()
    with pytest.raises(ValueError):
        setter(cls())
~~~

However, I think this is not readable and I'd leave the four single tests.

File: lib/matplotlib/axis.py (hunk 1 above)
Comment (timhoffm) [reply lib/matplotlib/axis.py]: I'd go for TypeError. By intent, the `hasattr` here is a kind of a loose type checking.

**Comment by jklymak** [comment]
@timhoffm while I have you: AttributeError, TypeError, or ValueError?

**Comment by timhoffm** [comment]
I'd go for TypeError. By intent, the hasattr here is a kind of a loose type checking.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -1568,6 +1568,9 @@ def set_major_formatter(self, formatter):
 
         ACCEPTS: A :class:`~matplotlib.ticker.Formatter` instance
         """
+        if not hasattr(formatter, 'format_data'):
+            raise TypeError("formatter argument should be instance of "
```
Comment (timhoffm) [inline lib/matplotlib/axis.py]: one could shorten this to 
~~~
raise TypeError("formatter must be a matplotlib.ticker.Formatter")
~~~
But that's a matter of taste.

Comment (jklymak) [reply lib/matplotlib/axis.py]: Thats what I had before but folks wanted "instance of" :wink:

**Comment by efiring** [comment]
I'm not asking for any more changes--it's good enough as-is--but I disagree with the switch to duck-typing here.  Duck-typing is a useful technique, but it is not always more appropriate than checking for membership in a class.  In the particular case here, there is no fundamental reason a Formatter needs a `format_data` method, although our base class has one; and the normal way of using the Formatter instance is via its '__call__', so someone could easily write a "Formatter" that would pass the quack-test used here but fail to work in practice.  Or one that would fail the quack test but would in fact work just fine. It is much less likely that a custom Formatter subclassed from our base would fail.  And the docstring specifies a Formatter, so testing for that is more consistent than quack-testing.

(It looks like we don't use the 'format_data' method anywhere except in the jpl_units test...)

**Comment by jklymak** [comment]
@efiring That actually makes sense to me.  It'd be pretty strange situation where someone bothered to pass a formatter or locator in that was not a subclass...  Further, its certainly bad to test on a method thats not used (`format_data`) or one thats too generic (`call`).  I'l wait a bit, maybe @timhoffm has a counter argument, but I'll make the suggested change before committing if not...

**Comment by story645** [comment]
While I'm with @timhoffm that I don't think it actually needs to be parameterized, pushing his example more:
```python
@pytest.mark.parametrize('setter', [ax.xaxis.set_minor_formatter, ax.xaxis.set_major_formatter])
@pytest.mark.parametrize('cls', [ matplotlib.ticker.LogFormatter, matplotlib.ticker.LogLocator])
def test_set_formatter_or_locator_type(setter, cls):
    fig, ax = plt.subplots()
    with pytest.raises(ValueError):
        setter(cls())
```
and agree with @efiring on `isinstance`.

**Comment by timhoffm** [comment]
Should this be backported?

**Comment by efiring** [comment]
The intention is to restrict backports to genuine bug-fixes, and soon only to *critical* bug-fixes, so this is not a candidate for backporting.
